fix(screen): missing search highlights when redrawing from timer

gelguy opened the pull request:
Fixes https://github.com/neovim/neovim/issues/14064.

I'm actually not sure of the effects of this PR: I reverted subsets of the changes in 910bbc3cca796f7fa941e0f6176cd0061de0e01c and removing these lines fixed the issue. Functional tests all pass so I assume nothing major breaks.

I wasn't able to write a lua test that fails either. I tried adding a test which tries the steps in #14064 to `functional/ui/searchhl_spec.lua` but the test doesn't fail before the changes are applied.
```vim
  it('correctly applies highlighting when redrawing in timer', function()
    feed_command('set hlsearch')
    feed_command('set incsearch')

    insert([[






      line 0
      line 1
      line 2
      line 3
      line 4
      line 5
      line 6
    ]])

    command("autocmd CmdlineChanged * call timer_start(0, {-> execute('redraw')})")

    feed("gg/0")
    screen:expect([[
                                              |
                                              |
                                              |
                                              |
                                              |
        line {3:0}                                |
      /0^                                      |
    ]])
  end)
```

IMO the whole patch can be reverted - I manually performed the actions in the Vim test and it doesn't fail even if the patch isn't applied.

gelguy commented:
Are there steps I can take to move this PR forward?

It's a major blocker for the plugin I'm working on.

justinmk commented:
can't argue with your reasoning. I see that https://github.com/neovim/neovim/pull/13074 doesn't have Lua screen tests for https://github.com/janlazo/neovim/commit/818e794f12171ab0fd62152e7d197da7bc43535f or https://github.com/janlazo/neovim/commit/910bbc3cca796f7fa941e0f6176cd0061de0e01c . 

[CheckRunVimInTerminal](https://github.com/janlazo/neovim/commit/910bbc3cca796f7fa941e0f6176cd0061de0e01c#diff-183e586cc07f732cf6aaf830dd0bc468dda6e77b099af6d5b341e88d2ea3a86aR280) always skips because Nvim doesn't support Vim's screen tests. 

Should be easy to write a Lua screen test that covers what `Test_cursor_column_in_concealed_line_after_window_scroll` does. Then we can revert https://github.com/neovim/neovim/commit/910bbc3cca796f7fa941e0f6176cd0061de0e01c which, yes, looks like typical crap fumbling around.

jityao committed:
Revert "vim-patch:8.1.2294: cursor pos wrong with concealing and search causes a scroll"

This reverts commit 910bbc3cca796f7fa941e0f6176cd0061de0e01c.

jityao committed:
Add lua version of Test_cursor_column_in_concealed_line_after_window_scroll

gelguy commented:
I've reverted the patch (but kept the Vim test) and added a lua functional test that should be identical. Note that the revert wasn't clean as the signatures of `update_topline()` and `win_update()` were changed.

There are lint errors but I've ignored them since I didn't want to change the reverted code too much.

justinmk commented on src/nvim/screen.c:
```diff
@@ -1745,51 +1743,18 @@ static void win_update(win_T *wp, Providers *providers)
-        }
       }
-      recursive = false;
+      recursive = FALSE;
```
```suggestion
      recursive = false;
```

justinmk commented on src/nvim/screen.c:
```diff
@@ -1745,51 +1743,18 @@ static void win_update(win_T *wp, Providers *providers)
-      const linenr_T old_topline = wp->w_topline;
-      const int new_wcol = wp->w_wcol;
-      recursive = true;
+      recursive = TRUE;
```
```suggestion
      recursive = true;
```

justinmk committed:
Update src/nvim/screen.c

justinmk committed:
Update src/nvim/screen.c

justinmk commented on src/nvim/screen.c:
```diff
@@ -763,10 +763,8 @@ static void win_update(win_T *wp, Providers *providers)
-  const linenr_T old_botline = wp->w_botline;
-  const int old_wrow = wp->w_wrow;
-  const int old_wcol = wp->w_wcol;
+  static int recursive = FALSE;         /* being called recursively */
```
```suggestion
  static bool recursive = false;  // being called recursively
```

justinmk committed:
Update src/nvim/screen.c

justinmk commented on src/nvim/screen.c:
```diff
@@ -764,9 +764,7 @@ static void win_update(win_T *wp, Providers *providers)
-  const linenr_T old_botline = wp->w_botline;
-  const int old_wrow = wp->w_wrow;
-  const int old_wcol = wp->w_wcol;
+  int old_botline = wp->w_botline;
```
don't see why the `const` or `linenr_T` declarations would be reverted. Basically, any non-functional style/conventional improvements should not be reverted.

justinmk committed:
Update src/nvim/screen.c

justinmk commented on src/nvim/screen.c:
```diff
@@ -1745,49 +1743,16 @@ static void win_update(win_T *wp, Providers *providers)
-      // New redraw either due to updated topline or due to wcol fix.
-      if (wp->w_redr_type != 0) {
-        // Don't update for changes in buffer again.
+      update_topline(curwin);         /* may invalidate w_botline again */
```
```suggestion
      update_topline(curwin);  // may invalidate w_botline again
```

justinmk committed:
Update src/nvim/screen.c

justinmk commented on src/nvim/screen.c:
```diff
@@ -1745,49 +1743,16 @@ static void win_update(win_T *wp, Providers *providers)
-      if (wp->w_redr_type != 0) {
-        // Don't update for changes in buffer again.
+      if (must_redraw != 0) {
+        /* Don't update for changes in buffer again. */
```
```suggestion
        // Don't update for changes in buffer again.
```

justinmk committed:
Update src/nvim/screen.c

justinmk commented on test/functional/ui/syntax_conceal_spec.lua:
```diff
@@ -913,4 +913,46 @@ describe('Screen', function()
+      screen:expect{grid=[[
+        setl conceallevel2 concealcursornc                   |
+        normal gg                                            |
+        "{5:^expr}                                                |
```
👍  The `^` here indicates that cursor is in column 2, which is what is expected: https://github.com/janlazo/neovim/commit/910bbc3cca796f7fa941e0f6176cd0061de0e01c#diff-183e586cc07f732cf6aaf830dd0bc468dda6e77b099af6d5b341e88d2ea3a86aR310

justinmk commented:
Let's try it. 

@janlazo  this adds a test which covers #13074 910bbc3cca796f7fa941e0f6176cd0061de0e01c while reverting the screen.c code changes from there. We can revert it if you see a problem.

zeertzjq commented:
> TODO: we also need a Lua screen test for https://github.com/janlazo/neovim/commit/818e794f12171ab0fd62152e7d197da7bc43535f

Added in #21353
